ci_tools/list_docs_tovalidate.py

```diff
@@ -0,0 +1,53 @@
+""" Script to list the objects that should be numpydoc validated
+"""
+import ast
+from _ast import FunctionDef, ClassDef, Module
+import argparse
+from git_evaluation_tools import get_diff_as_json as gdj
+
+parser = argparse.ArgumentParser(description='List the objects with docstrings in the files provided')
+parser.add_argument('gitdiff', metavar='gitdiff', type=str,
+                        help='Output of git diff between PR branch and base branch')
+parser.add_argument('output', metavar='output', type=str,
+                        help='File to save the output to')
+args = parser.parse_args()
+
+results = gdj(args.gitdiff)
+changes = {}
+for file, upds in results.items():
+    if file.startswith('pyccel/'):
+        for line_no in upds['addition']:
+            if file in changes:
+                changes[file].append(int(line_no))
+            else:
+                changes[file] = []
+                changes[file].append(int(line_no))
+
+objects = []
+for file, line_nos in changes.items():
+    with open(file,'r', encoding="utf-8") as f:
+        tree = ast.parse(f.read())
+    prefix = file[:-3].replace('/', '.')
+    for node in ast.walk(tree):
+        if isinstance(node, (FunctionDef, ClassDef, Module)):
+            if isinstance(node, Module):
+                objects.append(prefix)
+            if isinstance(node, FunctionDef):
+                if hasattr(node, "parent"):
+                    if any((node.lineno <= x <= node.end_lineno
+                            for x in line_nos)):
+                        objects.append('.'.join([prefix,node.parent.name, node.name]))
+                else:
+                    if any((node.lineno <= x <= node.end_lineno
+                            for x in line_nos)):
+                        objects.append('.'.join([prefix, node.name]))
+            if isinstance(node, ClassDef):
+                if any((node.lineno <= x <= node.end_lineno
+                        for x in line_nos)):
+                    objects.append('.'.join([prefix, node.name]))
+                for child in node.body:
+                    if isinstance(child, ast.FunctionDef):
+                        child.parent = node
```

Comment:
You are right! Yes, I am collecting all objects whose contents have been changed. The "parent" attribute here is used to differentiate between functions and class methods. The class name of a method is needed to get the full name of the method object. This will not take care of inner classes, but I don't think there is a need for that, is there?

Reply:
Not that I am aware of. I think we're safe for now. Although what about functions in functions?

Reply:
It is not handled either. I will refactor the code and mimic what you exposed in the following comment to take care of those cases as well. It is not hard to keep the absolute name of the object at all times

Reply:
Ah no, it should be handled thanks to walk. It just won't have a helpful name